relacy/thread.hpp

```diff
@@ -217,7 +217,7 @@ struct thread_info : thread_info_base
                     return (unsigned)-1; // access to unitialized var
 
                 history_t const& prev = var.history_[(index - 1) % atomic_history_size];
-                if (prev.busy_ && prev.last_seen_order_[index_] <= last_yield_)
+                if (prev.busy_ && prev.first_seen_order_[index_] <= last_yield_)

```

Comment:
This seems to make yields restrict more execution. Trying to convince myself that this is ok, but I can't because proving that some execution is useless for a checker is close to impossible (absence of evidence is not evidence of absence). What do you think?

The affected scenario is:
```
T1: X = 1
T2: r = X // observes 1
T2: YIELD
T2: r = X // observes 1
T1: X = 2
T2: r = X // previous we could yield 1 or 2, because YIELD happened before last_seen,
               // but now we will only yield 2, because YIELD happened after first_seen
```

Reply:
I see how the behavior will change in your example, but I actually don't understand what part of the memory model last_yield_ is supposed to address. What part of the model would be violated without respecting the last_yield_ ? Perhaps it is for some other reason, like guaranteeing forward progress in the history buffer without relying on randomness?

Reply:
This models 1.10/25:
```
An implementation should ensure that the last value (in modification order) assigned by an atomic or synchronization operation will become visible to all other threads in a finite period of time.
```

With YIELD being a practically useful way to detect time advancement in the formal model.

E.g. this will hang under exhaustive enumeration:
```
// thread 0
atomic_store(&X, 1, seq_cst);

// thread 1
while (!atomic_load(&X, seq_cst)) {}
```

But this won't:
```
// thread 0
atomic_store(&X, 1, seq_cst);

// thread 1
while (!atomic_load(&X, seq_cst)) { pthread_yield(); }
```

And that's exactly how people should write code, maybe not exactly that aggressive yielding, but overall there should be some yield ("I am waiting for something") point.

As I said, I don't see that your change is necessary negative. The simple example above will still obviously work. Probably we should commit it as is. It's just that there is a subtle change in behavior, and convincing either way is hard.

Reply:
Ah I see. Thanks for the explanation. I agree its difficult to show formally that this change will not matter. As you said, it seems unlikely to be a problem in practice because the difference manifests when a pair of loads that is _not_ interleaved with a yield follows a pair of loads that _is_ interleaved with a yield. This would be an usual pattern, but definitely possible.